rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -29,26 +29,79 @@ and we will not allow them to be provisioned. Although other resources might be
-they are not exposed through role bindings to any subjects (the cluster-admin role has internal Kubernetes bindings to "system:masters" group). 
-Currently, we only expose the cluster-admin role on GKE
-installations, since system:masters group is not available for impersonation there. We will expand on this and when
-discovering a cluster/installing kube-agent/creating a service account, we will
-create cluster role bindings for those roles, accordingly linking them to the groups "default-cluster-admin", "default-view" and
-"default-edit". It will give users an opportunity to always have standard set of Kubernetes permissions which they can use together
-with Teleport's fine-grained RBAC definitions for Kubernetes. This will make enrolling and starting using Kubernetes clusters
-into Teleport a much simpler process, since user will be able to setup RBAC completely in Teleport if they want. We will also be able
+### Improving Day 1 experience - Adding capability to manage Kubernetes RBAC from Teleport. 
+
+Currently users rely on `kubernetes_groups` and/or `kubernetes_users` fields of the role definition when setting up
+access for Kubernetes. This requires roles/bindings to be created by the client in Kubernetes cluster independently from the 
+RBAC setup in the Teleport. These roles/bindings also should be kept in sync with the Teleport RBAC side,
+which creates additional hurdles for clients, especially the new ones, when using Kubernetes access. 
+To improve this situation we will introduce auto-provisioning of the Kubernetes RBAC based on 
+the Teleport roles, which means users will be able to define required permissions completely in the Teleport, and under the hood
+we will keep them in sync transparently for the users.
+
+We will add a new field to the role definition - `kubernetes_permissions`. That field will contain desired Kubernetes RBAC permissions
+for that role and will be similar in structure to the native Kubernetes RBAC:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: kube-role
+spec:
+  allow:
+    kubernetes_labels:
+      "*": "*"
+    kubernetes_groups: ["kube_group"]
+    kubernetes_permissions:
+      namespaces: ["namespace1", "namespace2"]
+      rules:
+      - resources: ["pod", "pod/exec"]
+        verbs: ["get", "list", "create"]
+      - resources: ["deployments"]
+        apiGroups: ["app"]
+        verbs: ["*"]
+      - resources: ["secrets"]
+        apiGroups: [""]
+        resourceNames: ["secret1", "secret2"]
+        verbs: ["get", "list""]
+```
+
+Presence of this field will signal that user wants Teleport to auto-provision these permissions into the Kubernetes clusters.
+Teleport will translate permissions defined in that field into Kubernetes RBAC definitions and will automatically create roles and bindings
+on the target Kubernetes clusters (as specified by the labels). If `namespaces` field includes '*' Teleport will create ClusterRole/ClusterRoleBinding
```

Comment:
How do we avoid conflicts between two agents in the same cluster with identical role names? Hash the cluster name and prefix/suffix the kube RBAC resources?

This is definitely a thing people are doing and this will make adoption harder for them as they tend to have consistent role naming rules across their multiple Teleport clusters.

Reply:
You mean case when Kubernetes cluster is enrolled in two Teleport clusters? Yeah in that case we can a collision 🤔 
We can add prefix, so it would be something like `teleport:telecluster1:kuberole1`, I don't think we should just use cluster names instead of a hash. It does increase role names, unfortunately, but I guess it's not that bad since they are not supposed to be used manually by the users. I think it's still better for users if they are readable though.